man/threadFileLogger.3
.\" Written  Oct 2012 by Olivier Charloton (oliviers.open.source.softwares@gmail.com)
.\" heavily based on the work of Steve Greenland (stevegr@neosoft.com) on syslog man.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.TH THREADFILELOGGER 3 2012-10-06 "dbgflags" "Dbgflags Programmer's Manual"
.SH NAME
openLogThreadFile, closeLogThreadFile, threadFileLogger, vthreadFileLogger, setThreadFileLoggerMaxSize, setThreadFileLoggerMaxDuration, setThreadFileLoggerDirectory  \- log messages to files, one file per thread.
.SH SYNOPSIS
.B #include <dbgflags/loggers.h>
.sp
.BI "void openLogThreadFile(const char *" ident ", int " option ", int " facility ");"
.br
.BI "void threadFileLogger(int " priority ", const char *" format ", ...);"
.br
.BI "int setThreadFileLoggerMaxSize(const size_t " maxSizeInBytes ");"
.br
.BI "int setThreadFileLoggerMaxDuration(const time_t " maxDurationInSeconds ");"
.br
.BI "int setThreadFileLoggerDirectory(const char *" directory ");"
.br
.BI "void closeLogThreadFile(void);"
.sp
.B #include <stdarg.h>
.sp
.BI "void vthreadFileLogger(int " priority ", const char *" format ", va_list " ap ");"
.br
.SH DESCRIPTION
This set of functions offers the same operations than the fileLogger but use one file per thread.

.BR openLogThreadFile ()
create a new log file for a program's thread.  The string pointed
to by
.I ident
is prepended to every message, is typically set to the thread name and will be use to name the file (when set).
The
.I option
argument specifies flags which control the operation of
.BR openLogThreadFile ()
and subsequent calls to
.BR threadFileLogger ().
The
.I facility
argument establishes a default to be used if
none is specified in subsequent calls to
.BR threadFileLogger ().
Values for
.I option
and
.I facility
are given below.  The use of 
.BR openLogThreadFile ()
is optional; it will automatically be called by
.BR threadFileLogger ()
if necessary, in which case 
.I ident
will default to NULL.
.sp
.BR threadFileLogger ()
generates a log message, which will be written to a file.
The
.I priority
argument is formed by ORing the
.I facility
and the 
.I level
values (explained below).  The remaining arguments
are a 
.IR format ,
as in 
.BR printf (3)
and any arguments required by the 
.IR format , 
except that the two character sequence %m will be replaced by
the error message string 
.IR strerror ( errno ).
A trailing newline is added when needed.

The function
.BR vthreadFileLogger ()
performs the same task as
.BR threadFileLogger ()
with the difference that it takes a set of arguments which have
been obtained using the
.BR stdarg (3)
variable argument list macros.

.BR setThreadFileLoggerMaxSize()
set the maximum size in bytes of the log file of the 
.IR current
.IR thread.
When this size is reached a new log file will be created. 
Remark: messages are not splitted between log files so the log file's size will be greater or equal to the maximum size.

.BR setThreadFileLoggerMaxDuration()
set the maximum duration in seconds of the log file of the  
.IR current
.IR thread.
When this duration is reached a new log file will be created. 
Remark: messages are not splitted between log files so the log file's duration will be greater or equal to the maximum duration.

.BR setThreadFileLoggerDirectory()
set the directory where the log files will be located. Remarks: this parameter is common to all threads.

.BR closeLogThreadFile ()
closes the descriptor file being used to write to the log file.  The use of
.BR closeLogFile ()
is optional.

.SH PARAMETERS
This section lists the parameters used to set the values of 
.IR option , " facility" ", and " priority 
(sames as 
.BR syslog(3)
).
.SS option
The
.I option
argument to 
.BR openLogThreadFile ()
is an OR of any of these:
.TP
.B LOG_CONS
Write directly to system console if there is an error while logging to
a file.
.TP
.B LOG_NDELAY
Create the log file immediately (normally, the file is created when
the first message is logged).
.TP
.B LOG_NOWAIT
Ignored.
.TP
.B LOG_ODELAY
The converse of
.BR LOG_NDELAY ;
Ignored too.
.TP
.B LOG_PERROR
(Not in POSIX.1-2001.) Print to stderr as well.
.TP
.B LOG_PID
Include PID with each message.
.TP
.B LOG_TID
(dbgflags's extension) Include Thread's identity (TID) with each message.
.TP
.B LOG_RDTSC
(dbgflags's extension) Log the number of clock cycles, since the CPU was powered up or reset, with each message. 
Warning: this value is different between every core and CPU, when running on more then one, use the LOG_CLOCK instead.
.TP 
.B LOG_LEVEL
(dbgflags's extension) Log the message level as a string with each message.
.TP
.B LOG_CLOCK
(dbgflags's extension) Log a time, that represents monotonic time since some unspecified starting point, with each message.
.TP 
.B LOG_FILE_ROTATE
(dbgflags's extension to manage log files) Use two files to log messages, when the threshold (set by 
.BR setThreadFileLoggerMaxSize(3) 
or 
.BR setThreadFileLoggerMaxDuration(3)) 
is reached, 
the current log file is closed then renamed to old and a new one is created (the previous old file if any is deleted). 
.TP 
.B LOG_FILE_HISTO
(dbgflags's extension to manage log files) Each log file's name will contains the current start time (as  the  number of seconds since the Epoch, 1970-01-01 00:00:00 +0000 (UTC)) and previous files will not deleted when the defined threshold will be reached.
.TP 
.B LOG_FILE_DURATION
(dbgflags's extension to manage log files) Enable the log file swap on duration (default is on size).
.TP 
.B LOG_FILE_WITHOUT_SYNC
(dbgflags's extension to manage log files) Disabled the O_SYNC flag (synchronous I/O to avoid data lost in case of program crash) of the file creation parameters, see open(2) for details. Remark: this option has been added on request for performances reasons, but be aware that the last message written in the log file may  be no more the last one when set.
.TP 
.B LOG_FILE_SYNC_ON_ERRORS_ONLY
(dbgflags's extension to manage log files) Disabled the O_SYNC flag (synchronous I/O to avoid data lost in case of program crash) of the file creation parameters, see open(2) for details, but a data flush of the log file will be performed if the importance of the message is upper or equal to error. Remark: this flag offers a compromise with the previous one.

.SS facility 
The 
.I facility
argument is used to specify what type of program is logging the message.
This lets the configuration file specify that messages from different
facilities will be handled differently.
.TP
.B LOG_AUTH
security/authorization messages (DEPRECATED Use 
.B LOG_AUTHPRIV 
instead)
.TP
.B LOG_AUTHPRIV
security/authorization messages (private) 
.TP
.B LOG_CRON
clock daemon 
.RB ( cron " and " at )
.TP
.B LOG_DAEMON
system daemons without separate facility value
.TP
.B LOG_FTP
ftp daemon
.TP
.B LOG_KERN
kernel messages
.TP
.BR LOG_LOCAL0 " through " LOG_LOCAL7
reserved for local use
.TP
.B LOG_LPR
line printer subsystem
.TP
.B LOG_MAIL
mail subsystem
.TP
.B LOG_NEWS
USENET news subsystem
.TP
.B LOG_SYSLOG
messages generated internally by
.B syslogd 
.TP
.BR LOG_USER " (default)"
generic user-level messages 
.TP
.B LOG_UUCP
UUCP subsystem

.SS level
This determines the importance of the message.  The levels are, in order
of decreasing importance:
.TP
.B LOG_EMERG
system is unusable
.TP
.B LOG_ALERT
action must be taken immediately
.TP
.B LOG_CRIT
critical conditions
.TP
.B LOG_ERR
error conditions
.TP
.B LOG_WARNING
warning conditions
.TP
.B LOG_NOTICE
normal, but significant, condition
.TP
.B LOG_INFO
informational message
.TP
.B LOG_DEBUG
debug-level message
.LP
The functions
.BR setlogmaskex (3)
or
.BR setlogmask (3)
can be used to restrict logging to specified levels only.
.SH NOTES
Never pass a string with user-supplied data as a format, use the folloing instead:
.RS
threadFileLogger(priority, "%s", string);

.SH "SEE ALSO"
.BR logger (1),
.BR syslog (3),
.BR syslogex (3),
.BR fileLogger (3),
.BR setlogmask (3),



NEWS
- A few corrections to build on a Debian 6 64bits.
- pkgconfig test program build corrected when the library hase not been already installed.
- the gettid internal function use the syscall function (instead of the "int 0x80") when building on a 2.6.x Linux kernel.
- Buffer overflow correction in consoleLogger & fileLogger (use a stream buffer now instead of a buffer on the stack).
  and new test programs to check the library's logger with message of size up to 4096 bytes.
- thread cancellation points management in fileLogger added (useless in the consoleLogger).
- added instructions to free the used buffer in the thread cancellation points management of the syslogex logger (glibc bug ?)
- logger sending trace to one file per thtread

Remarques:
- Files'repository moved to GIT and "public" version published to github
- using redmine to manage evolutions (ER) and anomalies reports (AR).


libdbgflags/threadFileLogger.c
#include "config.h"
#include <stdio.h>
#include <stdarg.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <pthread.h>
#include <time.h>
#include <fcntl.h>
#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <linux/limits.h>
#include <dbgflags/syslogex.h>
#include <dbgflags/dbgflags.h>
#include <dbgflags/loggers.h>
#include "system.h"

#define DEBUG_LOG_HEADER "threadFileLogger"
#define LOGGER consoleLogger
#define LOG_OPTS  0
#define MAX_LOGTAG_SIZE	128
#define LOG_EXTENSION 		".log"
#define BAK_LOG_EXTENTION 	".bak"
#define LONGER_EXTENSION	BAK_LOG_EXTENTION

#ifdef _DEBUGFLAGS_H_
static DebugFlags debugFlags = {
		"threadFileLogger",
		{
				"threadFile"
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""
				,""}
		,0xFFFFFFFF
};
#define FLAG_THREAD_FILE  FLAG_1

#else /* _DEBUGFLAGS_H_ */
#define FILTER
#define FLAG_FILE
#endif /* _DEBUGFLAGS_H_ */

#include <dbgflags/debug_macros.h>

#define MODULE_FLAG FLAG_THREAD_FILE

#ifdef _DEBUG_
#include <assert.h>
#define ASSERT assert
#else
#define ASSERT(x)
#endif /* _DEBUG_ */

static char fullProcessName[PATH_MAX];
static char directory[PATH_MAX];
static char *processName = NULL;
static char *fileDirectory = NULL;
//static pthread_mutex_t fileLock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;

//static int LogMask = 0xff; /* mask of priorities to be logged */

#include "filesUtils.h"
#define LOG_TAG_USED	(1<<0) /* set when the LogTag has been used for the first time to build the file's name */

typedef struct ThreadFileData_ {
	int	LogStat;
	int	LogFacility;
	char fullFileName[PATH_MAX];
	/*char directory[PATH_MAX];  useless */
	char LogTag[MAX_LOGTAG_SIZE];		/* string to tag the entry with */
	size_t maxSize; /* in bytes */
	time_t maxDuration; /* in seconds */
	int logFile; /* file's descriptor */
	size_t fileSize; /* in bytes */
	time_t startTime; /* UNIX EPOCH time */
	unsigned int miscFlags;
#ifdef FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD
	size_t currentPageCacheMaxSize; /* max size of the current file system page cache */
#endif /* FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD */
} ThreadFileData;

static inline void initThreadFileData(ThreadFileData *data) {
	/* set default values */
	data->LogStat = 0; /* status bits, set by openlog() */
	data->LogFacility = LOG_USER; /* default facility code */
	data->fullFileName[0] = '\0'; /* will be set by the createfile function */
	/*data->directory[0] = '\0';*/ /* will be set by the createfile function or the openLogThreadFile function */
	data->LogTag[0] = '\0';
	data->maxSize = (1024*100); /* 100 ko */
	data->maxDuration = 24*60*60; /* 24 hours */
	data->logFile = -1;
	data->fileSize = 0;
	data->startTime = 0;
	data->miscFlags = 0x0;
#ifdef FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD
	data->currentPageCacheMaxSize = 0;
#endif /* FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD */
}

/* TSD key used to store information used to manage on file per thread */
static pthread_key_t TLSThreadFileKey;

/* Once-only initialisation of the key */
static pthread_once_t TLSThreadFileKeyOnce = PTHREAD_ONCE_INIT;

static void freeThreadFileData(void * buffer) {  
	if (buffer) {
		ThreadFileData *threadFileData = (ThreadFileData *)buffer;
		if (threadFileData->logFile != -1) {
			close(threadFileData->logFile);
			threadFileData->logFile = -1;
		}
		free(buffer);
	}
}

static void initThreadFileLoggers() {
	pthread_key_create(&TLSThreadFileKey, freeThreadFileData);
	setProcessName();
}

static inline ThreadFileData *getThreadFileData() {
	pthread_once(&TLSThreadFileKeyOnce, initThreadFileLoggers);
	ThreadFileData *threadFileData = (ThreadFileData *)pthread_getspecific(TLSThreadFileKey);
	if (unlikely(NULL == threadFileData)) {
		/* first call for this thread:
		 * allocate memory to store information for it and initialize them
		 */
		threadFileData = malloc(sizeof(ThreadFileData));
		if (likely(threadFileData != NULL)) {
			initThreadFileData(threadFileData);
			pthread_setspecific(TLSThreadFileKey, threadFileData);
		} else {
			const pid_t tid = gettid();
			ERROR_MSG("failed to allocate %u bytes to manage the log file for the thread %u",sizeof(ThreadFileData),tid);
		}
	} //else already set
	return threadFileData;
}

static __inline int renamePreviousLogFiles(ThreadFileData *threadFileData) {
	int error = EXIT_SUCCESS;
	DIR *logdir = opendir(directory);
	if (logdir != NULL) {
		int directoryfd = dirfd(logdir);
		if (directoryfd != -1) {
			struct dirent64 *directoryEntry = NULL;
			char baseFileNameBuffer[PATH_MAX];
			const pid_t tid = gettid();
			const int n = sprintf(baseFileNameBuffer,"%s",processName);
			while ((directoryEntry = readdir64(logdir)) != NULL) {
				if ((DT_REG == directoryEntry->d_type) && (strncmp(baseFileNameBuffer,directoryEntry->d_name,n) == 0)) {
					/* append the tag name just before the file's extension */
					char newFileNameBuffer[PATH_MAX];
					char *extensionStart = NULL;
					char extension[5];
					const char *oldFileNameBuffer = directoryEntry->d_name;
					DEBUG_MSG("%s matchs %s",directoryEntry->d_name,baseFileNameBuffer);
					strcpy(newFileNameBuffer,oldFileNameBuffer);
					extensionStart = strrchr(newFileNameBuffer,'.');
					if (extensionStart) {
						strncpy(extension,extensionStart+1,sizeof(extension)-1);
						extension[sizeof(extension)-1] = '\0';
						sprintf(extensionStart,"_%s.%s",threadFileData->LogTag,extension);
						DEBUG_VAR(newFileNameBuffer,"%s");
						if (unlikely(renameat(directoryfd,oldFileNameBuffer,directoryfd,newFileNameBuffer) == -1)) {
							error = errno;
							ERROR_MSG("rename %s => %s error %d (%m)", oldFileNameBuffer, newFileNameBuffer, error);
						}
					} else {
						error = ENOENT;
						ERROR_MSG("extension not found in %d",newFileNameBuffer);
					}
				}
			} //while ((directoryEntry = readdir64(logdir)) != NULL)
		} else {
			error = errno;
			ERROR_MSG("dirfd on DIR %s error %d (%m)",directory,error);
		}
		closedir(logdir);
		logdir = NULL;
	} else {
		error = errno;
		ERROR_MSG("opendir %s error %d (%m)",directory,error);
	}
	return error;
}

static __inline int setThreadFullFileName(ThreadFileData *threadFileData) {
	int error = EXIT_SUCCESS;
	ASSERT(threadFileData);
	const pid_t tid = gettid();
	const int LogStat = threadFileData->LogStat;
	char *fullFileName = threadFileData->fullFileName;

	if (LOG_FILE_ROTATE & LogStat) {
		/*
		 * naming is done according to the following pattern:
		 * directory / processName_<tid>_<LogTag> [.log|.bak]
		 */
		char oldFileNameBuffer[PATH_MAX];
		size_t n = strlen(threadFileData->fullFileName);
		if (unlikely(0 == n)) {
			/* first call, currently we don't manage the case the logger is used BEFORE the LogTag is set */
			if (threadFileData->LogTag[0]) {
				if (unlikely(!(threadFileData->miscFlags & LOG_TAG_USED))) {
					// first call with a LogTag set: try to rename all previously logfiles (if any)
					error = renamePreviousLogFiles(threadFileData);
					if (EXIT_SUCCESS == error) {
						threadFileData->miscFlags |= LOG_TAG_USED;
					}
				}
				sprintf(oldFileNameBuffer,"%s/%s_%u_%s"BAK_LOG_EXTENTION, directory, processName,tid,threadFileData->LogTag);
				sprintf(fullFileName,"%s/%s_%u_%s"LOG_EXTENSION, directory, processName,tid,threadFileData->LogTag);
			} else {
				sprintf(oldFileNameBuffer, "%s/%s_%u"BAK_LOG_EXTENTION, directory, processName,tid);
				sprintf(fullFileName,"%s/%s_%u"LOG_EXTENSION, directory, processName,tid);
			}
		} else {
			/* compute the oldFileNameBuffer from the current file's name */
			const size_t n = strlen(fullFileName);
			char *suffixe = NULL;
			strcpy(oldFileNameBuffer,fullFileName);
			suffixe = oldFileNameBuffer + n - strlen(LOG_EXTENSION); /* strlen will be computed during the compilation */
			strcpy(suffixe,BAK_LOG_EXTENTION);
		}

		if (unlikely(unlink(oldFileNameBuffer) == -1)) {
			error = errno;
			if (error != ENOENT) {
				ERROR_MSG("unlink %s error %d (%m)", oldFileNameBuffer, error);
			}
		}

		if (unlikely(rename(fullFileName, oldFileNameBuffer) == -1)) {
			error = errno;
			if (error != ENOENT) {
				ERROR_MSG("rename %s => %s error %d (%m)", fullFileName,oldFileNameBuffer, error);
			}
		}

	} else if (LOG_FILE_HISTO & LogStat) {
		/*
		 * naming is done according to the following pattern:
		 * directory / processName_<pid>_<tid>_<LogTag>_<time>.log
		 */
		const pid_t pid = getpid();
		time_t currentTime = time(NULL);
		if (unlikely(((time_t) - 1) == currentTime)) {
			currentTime = 0;
		}
		if (threadFileData->LogTag[0]) {
			sprintf(fullFileName, "%s/%s_%u_%u_%s_%u"LOG_EXTENSION, directory, processName,pid,tid,threadFileData->LogTag,currentTime);
		} else {
			sprintf(fullFileName, "%s/%s_%u_%u_%u"LOG_EXTENSION, directory, processName,pid,tid,currentTime);
		}
	} else {
		sprintf(fullFileName, "%s/%s_%u"LOG_EXTENSION, directory, processName,tid);
	}
	DEBUG_VAR(fullFileName, "%s");
	return error;
}

static inline int createFile(ThreadFileData *threadFileData) {

#ifdef _DEBUGFLAGS_H_
	{
		static unsigned int registered = 0;
		if (unlikely(0 == registered)) {
			registered = 1; /* dirty work around to avoid deadlock: syslogex->register->syslogex */
			registered = (registerLibraryDebugFlags(&debugFlags) == EXIT_SUCCESS);
		}
	}
#endif /*_DEBUGFLAGS_H_*/
	int error = EXIT_SUCCESS;
	ASSERT(threadFileData);

	int internalError = EXIT_SUCCESS;
	int flags = O_WRONLY|O_CREAT|O_TRUNC;
	const int LogStat = threadFileData->LogStat;

	if (likely((!(LogStat & LOG_FILE_WITHOUT_SYNC)) && (!(LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY)))) {
		flags |= O_SYNC; /* enable synchronous I/O to avoid data lost in case of crash */
	}

	if (threadFileData->logFile != -1) {
		WARNING_MSG("logFile was NOT NULL");
		if (close(threadFileData->logFile) != 0) {
			internalError = errno;
			ERROR_MSG("close %d error %d (%m)",threadFileData->logFile, internalError);
		}
		threadFileData->logFile = -1;
	}

	setThreadFullFileName(threadFileData);

	threadFileData->logFile = open(threadFileData->fullFileName,flags,S_IRUSR|S_IWUSR|S_IRGRP);
	if (likely(threadFileData->logFile != -1)) {
		threadFileData->fileSize = 0;
		if (LOG_FILE_DURATION & threadFileData->LogStat) {
			threadFileData->startTime = time(NULL);
		}
	} else {
		error = errno;
		ERROR_MSG("open %s for creation error %d (%m)",threadFileData->fullFileName, error);
	}

	return error;
}

void openLogThreadFile(const char *ident, int logstat, int logfac) {
	ThreadFileData *threadFileData = getThreadFileData();
	if (likely(threadFileData)) {
		if (ident != NULL) {
			strncpy(threadFileData->LogTag,ident,MAX_LOGTAG_SIZE);
			threadFileData->LogTag[MAX_LOGTAG_SIZE-1] = '\0';
		}
		threadFileData->LogStat = logstat;
		if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0) {
			threadFileData->LogFacility = logfac;
		}
	} //error already printed
}

typedef struct cleanup_args_ {
	void *buffer;
	//char *logBuffer;
	/* mutex is a static global variable so... */
} cleanup_args;

static void cancel_handler(void *ptr) {
	cleanup_args *args = (cleanup_args *)ptr;
	if ((args != NULL)){
		if (args->buffer != NULL) {
			free(args->buffer);
			args->buffer = NULL;
		}
	}
}

/* strcpy then strcat function using the same cursor */
static inline void strcpyNcat(char *buffer, const char *source1,const char *source2) {
	register char *p = buffer;

	register const char *s = source1;
	while(*s != '\0') {
		*p = *s;
		p++;
		s++;
	}

	s = source2;
	while(*s != '\0') {
		*p = *s;
		p++;
		s++;
	}

	*p = '\0';
}

void vthreadFileLogger(int priority, const char *format, va_list optional_arguments) {
	int n = 0;
	int error = EXIT_SUCCESS;
	const int saved_errno = errno;
	int internalError = EXIT_SUCCESS;
	const int LogMask = setlogmask(0);

	/* Check priority against setlogmask values. */
	if ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) {
		ThreadFileData *threadFileData = getThreadFileData();
		if (likely(threadFileData)) {
			char *buf = 0;
			size_t bufsize = 1024;
			FILE *f = open_memstream(&buf, &bufsize);
			if (f != NULL) {
				struct tm now_tm;
				time_t now;
				const int LogStat = threadFileData->LogStat;

				(void) time(&now);
				n = strftime(f->_IO_write_ptr,f->_IO_write_end - f->_IO_write_ptr,"%h %e %T ",localtime_r(&now, &now_tm));
				f->_IO_write_ptr += n;

				if (threadFileData->LogTag[0]) {
					n += fprintf(f,"%s: ",threadFileData->LogTag);
				}

				if (LogStat & LOG_PID) {
					if (LogStat & LOG_TID) {
						const pid_t tid = gettid();
						n +=fprintf(f,"[%d:%d]", (int) getpid (),(int) tid);
					} else {
						n += fprintf(f,"[%d]", (int) getpid ());
					}
				} /* (LogStat & LOG_PID) */

				if (LogStat & LOG_RDTSC) {
					const unsigned long long int  t = rdtsc();
					n += fprintf(f,"(%llu)",t);
				} /* (LogStat & LOG_RDTSC) */

				if (LogStat & LOG_CLOCK) {
#if HAVE_CLOCK_GETTIME
					struct timespec timeStamp;
					if (clock_gettime(CLOCK_MONOTONIC,&timeStamp) == 0) {
						/*cursor += sprintf (cursor, "(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);*/
						n += fprintf(f,"(%lu.%.9d)",timeStamp.tv_sec,timeStamp.tv_nsec);
					} else {
						const int error = errno;
						ERROR_MSG("clock_gettime CLOCK_MONOTONIC error %d (%m)",error);
					}
#else
					static unsigned int alreadyPrinted = 0; /* to avoid to print this error msg on each call */
					if (unlikely(0 == alreadyPrinted)) {
						ERROR_MSG("clock_gettime  not available on this system");
						alreadyPrinted = 1;
					}
#endif /* HAVE_CLOCK_GETTIME */
				} /* (LogStat & LOG_CLOCK) */

				if (LogStat & LOG_LEVEL) {
					switch(LOG_PRI(priority)) {
					case LOG_EMERG:
						n += fprintf(f, "[EMERG]");
						break;
					case LOG_ALERT:
						n += fprintf(f, "[ALERT]");
						break;
					case LOG_CRIT:
						n += fprintf(f, "[CRIT]");
						break;
					case LOG_ERR:
						n += fprintf(f, "[ERROR]");
						break;
					case LOG_WARNING:
						n += fprintf(f, "[WARNING]");
						break;
					case LOG_NOTICE:
						n += fprintf(f, "[NOTICE]");
						break;
					case LOG_INFO:
						n += fprintf(f, "[INFO]");
						break;
					case LOG_DEBUG:
						n += fprintf(f, "[DEBUG]");
						break;
					default:
						fprintf(f,"[<%d>]",priority);
					} /* switch(priority) */
				} /* (LogStat & LOG_LEVEL) */

				errno = saved_errno; /* restore errno for %m format */
				n += vfprintf(f, format, optional_arguments);

				/* Close the memory stream; this will finalize the data
		into a malloc'd buffer in BUF.  */
				fclose(f);

				/*
				 * Some of the function used below are thread cancellation points
				 * (according to Advanced Programming in the Unix Environment 2nd ed p411)
				 * so set the handler to avoid deadlocks and memory leaks
				 */
				cleanup_args cancelArgs;
				cancelArgs.buffer = buf;
				pthread_cleanup_push(cancel_handler, &cancelArgs);

				if (LogStat & LOG_PERROR) {
					/* add the format argument to the header */
					const size_t n = strlen(format);
					const size_t size = bufsize + n + 1;
					char *logFormatBuffer = (char *)malloc(size);
					if (logFormatBuffer) {
						pthread_cleanup_push(free,logFormatBuffer);
						strcpyNcat(logFormatBuffer,buf,format);
						vfprintf(stderr, logFormatBuffer, optional_arguments);
						pthread_cleanup_pop(1); /* pop the handler and free the allocated memory */
					} else {
						ERROR_MSG("failed to allocate %u bytes to print the msg on stderr",size);
					}
				} /* (LogStat & LOG_PERROR) */

				/* file management */
				if (LOG_FILE_DURATION & LogStat) {
					const time_t currentTime = time(NULL);
					const time_t elapsedTime = currentTime - threadFileData->startTime;
					if (unlikely(elapsedTime >= threadFileData->maxDuration)) {
						close(threadFileData->logFile);
						threadFileData->logFile = -1;
					}
				} /* (LOG_FILE_DURATION & LogStat) */

				if (unlikely(-1 == threadFileData->logFile)) {
					error = createFile(threadFileData);
					/* error already logged */
				}

				if (likely(EXIT_SUCCESS == error)) {
					const ssize_t written = write(threadFileData->logFile,buf,n);
					if (written > 0) {
						if (unlikely(written != n)) {
							ERROR_MSG("only %d byte(s) of %d has been written to %s",written,n,threadFileData->fullFileName);
							if (LogStat & LOG_CONS) {
								int fd = open("/dev/console", O_WRONLY | O_NOCTTY, 0);
								if (fd >= 0 ) {
									dprintf(fd,"logMsg");
									close(fd);
									fd = -1;
								}
							} /* (LogStat & LOG_CONS) */
						} /* (unlikely(written != n)) */

						if (unlikely((LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY) && (LOG_PRI(priority) <= LOG_ERR))) {
							/* flush data if the log priority is "upper" or equal to error */
							error = fdatasync(threadFileData->logFile);
							if (error != 0) {
								error = errno;
								ERROR_MSG("fdatasync to %s error %d (%m)",threadFileData->fullFileName,error);
							}
							error = posix_fadvise(threadFileData->logFile, 0,0,POSIX_FADV_DONTNEED); /* tell the OS that log message bytes could be released from the file system cache */
							if (error != 0) {
							  ERROR_MSG("posix_fadvise to %s error %d (%m)",threadFileData->fullFileName,error);
							}
						} /* (unlikely((LogStat & LOG_FILE_SYNC_ON_ERRORS_ONLY) && (LOG_PRI(priority) <= LOG_ERR))) */
						threadFileData->fileSize += written;

#ifdef FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD
						threadFileData->currentPageCacheMaxSize += written;
						if (threadFileData->currentPageCacheMaxSize >= FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD) {
							/* tell the OS that log message bytes could be released from the file system cache */
							if (likely(posix_fadvise(threadFileData->logFile, 0,0,POSIX_FADV_DONTNEED) == 0)) {
								threadFileData->currentPageCacheMaxSize = 0;
								DEBUG_MSG("used file system cache allowed to be flushed (size was %u)",threadFileData->currentPageCacheMaxSize);
							} else {
								NOTICE_MSG("posix_fadvise to %s error %d (%m), current page cache max size is %u",threadFileData->fullFileName,error,threadFileData->currentPageCacheMaxSize);
							}
						}
#endif /* FILESYSTEM_PAGE_CACHE_FLUSH_THRESHOLD */

						if ((LOG_FILE_ROTATE & LogStat) || (LOG_FILE_HISTO & LogStat)) {
							if (threadFileData->fileSize >= threadFileData->maxSize) {
								close(threadFileData->logFile);
								threadFileData->logFile = -1;
							}
						}
					} else if (0 == written) {
						WARNING_MSG("nothing has been written in %s", threadFileData->fullFileName);
					} else {
						error = errno;
						ERROR_MSG("write to %s error %d (%m)", threadFileData->fullFileName, error);
					}
				} /*(likely(EXIT_SUCCESS == error)) */

				pthread_cleanup_pop(0);  /* moved to avoid a build error, use the preprocessor to understand why
				 * or have a look on man 3 pthread_cleanup_push:
				 * push & pop MUST BE in the SAME lexical nesting level !!!
				 */
				free(buf);
				cancelArgs.buffer = buf = NULL;
			} else { /* open_memstream(&buf, &bufsize) == NULL */
				ERROR_MSG("failed to get stream buffer");
				/* TMP! display the raw message to the console */
				vfprintf(stderr,format,optional_arguments);
				/* TODO: write the raw message to the file */
			}
		} /* (likely(threadFileData)) error already logged */
	} /* ((LOG_MASK (LOG_PRI (priority)) & LogMask) != 0) */
	/* message has not to be displayed because of the current LogMask and its priority */

	/*return n;*/
}

void threadFileLogger(int priority, const char *format, ...) {
	va_list optional_arguments;
	va_start(optional_arguments, format);
	vthreadFileLogger(priority, format,  optional_arguments);
	va_end(optional_arguments);
}

int setThreadFileLoggerMaxSize(const size_t maxSizeInBytes) {
	int maxSize = 0;
	ThreadFileData *threadFileData = getThreadFileData();
	if (likely(threadFileData)) {
		if (maxSizeInBytes != 0) {
			threadFileData->maxSize = maxSizeInBytes;
		}
		maxSize = threadFileData->maxSize;
	} else { /* error already logged */
		maxSize = -ENOMEM;
	}
	return maxSize;
}

int setThreadFileLoggerMaxDuration(const time_t maxDurationInSeconds) {
	int maxDuration = 0;
	ThreadFileData *threadFileData = getThreadFileData();
	if (likely(threadFileData)) {
		if (maxDurationInSeconds != 0) {
			threadFileData->maxDuration = maxDurationInSeconds;
		}
		maxDuration = threadFileData->maxDuration;
	} else { /* error already logged */
		maxDuration = -ENOMEM;
	}
	return maxDuration;
}

int setThreadFileLoggerDirectory(const char *logfileDirectory) {
	int error = EINVAL;
	ThreadFileData *threadFileData = getThreadFileData();
	if (likely(threadFileData)) {
		if (logfileDirectory != NULL) {
			if (logfileDirectory[0] != '$') {
				error = checkDirectory(logfileDirectory);
			} else {
				char *envVar = getenv(logfileDirectory+1);
				if (envVar != NULL) {
					error = checkDirectory(logfileDirectory);
				} else {
					error = ENOENT;
					ERROR_MSG("environment variable %s is not found",logfileDirectory);
				}
			} /* !(logfileDirectory[0] != '$') */
			if (EXIT_SUCCESS == error) {
				strcpy(directory, logfileDirectory);
				DEBUG_VAR(directory, "%s");
			}
		} /* (logfileDirectory != NULL) */
	} else { /* error already logged */
		error = ENOMEM;
	}

	return error;
}

void closeLogThreadFile(void) {
	ThreadFileData *threadFileData = getThreadFileData();
	if (likely(threadFileData)) {
		if (threadFileData->logFile != -1) {
			close(threadFileData->logFile);
			threadFileData->logFile = -1;
		}
	}
}

#include <dbgflags/ModuleVersionInfo.h>
MODULE_NAME(threadFileLogger);
MODULE_AUTHOR(Olivier Charloton);
MODULE_VERSION(1.0);
MODULE_FILE_VERSION(1.0);
MODULE_DESCRIPTION(logger to files (one file per thread) with the same signature than syslog using POSIX API);
MODULE_COPYRIGHT(LGPL);
